shark-ai/export_and_serve.py

```diff
@@ -0,0 +1,412 @@
+            logger.info(f"TP8 model config will be saved to {self.model_config_path}")
+
+        if self.compile:
+            logger.info("Compilation is enabled")
+            if self.ir_path is not None:
+                logger.info(f"IR for compilation located at {self.ir_path}")
+                self.mlir_path = self.ir_path
+
+        logger.info(f"Exported model will be saved to {self.export_dir}")
+
+    def export_tp1(self) -> None:
+        batch_size_string = ",".join([str(b) for b in self.export_batch_sizes])
+        logger.info(f"Exporting with batch sizes {batch_size_string}")
+
+        # TODO(vinayakdsci): Refactor this into a separate function.
+        export_subp = subprocess.run(
+            [
+                "python3",
+                "-m",
+                "sharktank.examples.export_paged_llm_v1",
+                f"--gguf-file={self.weight_loc}",
+                f"--output-mlir={str(self.mlir_path)}",
+                f"--output-config={self.model_config_path}",
+                f"--bs={batch_size_string}",
+            ],
+            capture_output=True,
+        )
+
+        # Exit if the command fails, no point in throwing an exception.
+        if export_subp.returncode != 0:
+            logger.error("Failed to export model in TP1 mode")
+            print(export_subp.stderr.decode(), file=sys.stderr)
+            exit(export_subp.returncode)
+
+        logger.info(
+            f"Successfully exported and saved model MLIR and config to {self.export_dir}"
+        )
+
+    def export_tp8(self) -> None:
+        shard_file = os.path.join(
+            str(self.artifacts_dir),
+            str(self.weight_loc).split("/")[-1][:-5] + "_tp8.irpa",
+        )
+        shard_subp = subprocess.run(
+            [
+                "python3",
+                "-m",
+                "sharktank.examples.sharding.shard_llm_dataset",
+                f"--irpa-file={self.weight_loc}",
+                f"--output-irpa={shard_file}",
+                "--tensor-parallelism-size=8",
+            ],
+            capture_output=True,
+        )
+
+        if shard_subp.returncode != 0:
+            logger.error("Failed to shard model to TP8")
+            print(shard_subp.stderr.decode(), file=sys.stderr)
+            exit(shard_subp.returncode)
+
+        logger.info(
+            f"Successfully sharded weight file; Unranked weight file located at {shard_file}"
+        )
+
+        # TODO(vinayakdsci): Refactor this into a separate function.
+        logger.info("Starting TP8 model export")
+        export_subp = subprocess.run(
+            [
+                "python3",
+                "-m",
+                "sharktank.examples.export_paged_llm_v1",
+                f"--irpa-file={shard_file}",
+                f"--output-mlir={str(self.mlir_path)}",
+                f"--output-config={self.model_config_path}",
+                "--bs=4",
+            ],
+            capture_output=True,
+        )
```

Comment:
Prefer to call the Python functions directly in-process instead of using `subprocess`. That way you get integrated error handling and ensure that the same python interpreter is used (this could also use `sys.executable` instead of `python3`